### ISSUE TITLE
PAM errors/warnings when starting a user session

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#28514]
Split out of #28504.

I'm going through logs and trying to reproduce this, and while doing so I noticed a couple of warnings/errors (some of them we even discussed on Telegram iirc):

```
[   29.173407] H testsuite-55.sh[664]: + systemctl --machine testuser@.host --user status testsuite-55-testbloat.service
[   29.210845] H systemd[1]: Started run-u50.service.
[   29.224695] H systemd[1]: Started session-12.scope.
[   29.226876] H (o-bridge)[893]: pam_unix(login:session): session opened for user testuser(uid=4711) by (uid=0)
[   29.234523] H testsuite-55.sh[891]: × testsuite-55-testbloat.service - Create a lot of memory pressure
[   29.234523] H testsuite-55.sh[891]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testbloat.service; static)
[   29.234523] H testsuite-55.sh[891]:      Active: failed (Result: oom-kill) since Mon 2023-07-24 22:05:47 CEST; 950ms ago
[   29.234523] H testsuite-55.sh[891]:    Duration: 13.720s
[   29.234523] H testsuite-55.sh[891]:     Process: 787 ExecStart=stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1 (code=killed, signal=KILL)
[   29.234523] H testsuite-55.sh[891]:    Main PID: 787 (code=killed, signal=KILL)
[   29.234523] H testsuite-55.sh[891]:         CPU: 270ms
[   29.246102] H dbus-broker[777]: Peer :1.12 is being disconnected as it sent a message with an invalid header.
[   29.246429] H (sd-pam)[894]: pam_systemd(login:session): Failed to release session: Access denied
[   29.246720] H testsuite-55.sh[664]: + break
[   29.246720] H testsuite-55.sh[664]: + systemctl --machine testuser@.host --user status testsuite-55-testbloat.service
[   29.249797] H (sd-pam)[894]: pam_unix(login:session): session closed for user testuser
[   29.249949] H (sd-pam)[894]: PAM Attempted to close sd-bus after fork, this should not happen.
[   29.250586] H systemd[1]: run-u50.service: Deactivated successfully.
[   29.250866] H dbus-broker[650]: A security policy denied :1.52 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
[   29.254817] H systemd[1]: session-12.scope: Deactivated successfully.
[   29.279868] H systemd[1]: Started run-u53.service.
[   29.295880] H systemd[1]: Started session-13.scope.
[   29.303401] H (o-bridge)[900]: pam_unix(login:session): session opened for user testuser(uid=4711) by (uid=0)
[   29.313469] H dbus-broker[777]: Peer :1.13 is being disconnected as it sent a message with an invalid header.
[   29.313728] H testsuite-55.sh[895]: × testsuite-55-testbloat.service - Create a lot of memory pressure
[   29.313728] H testsuite-55.sh[895]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testbloat.service; static)
[   29.313728] H testsuite-55.sh[895]:      Active: failed (Result: oom-kill) since Mon 2023-07-24 22:05:47 CEST; 1s ago
[   29.313728] H testsuite-55.sh[895]:    Duration: 13.720s
[   29.313728] H testsuite-55.sh[895]:     Process: 787 ExecStart=stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1 (code=killed, signal=KILL)
[   29.313728] H testsuite-55.sh[895]:    Main PID: 787 (code=killed, signal=KILL)
[   29.313728] H testsuite-55.sh[895]:         CPU: 270ms
[   29.323777] H (sd-pam)[901]: pam_systemd(login:session): Failed to release session: Access denied
[   29.324036] H testsuite-55.sh[664]: + systemctl --machine testuser@.host --user status testsuite-55-testchill.service
[   29.325619] H (sd-pam)[901]: pam_unix(login:session): session closed for user testuser
[   29.325743] H (sd-pam)[901]: PAM Attempted to close sd-bus after fork, this should not happen.
[   29.326192] H systemd[1]: run-u53.service: Deactivated successfully.
[   29.326425] H dbus-broker[650]: A security policy denied :1.55 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
[   29.331144] H systemd[1]: session-13.scope: Deactivated successfully.
[   29.353609] H systemd[1]: Started run-u56.service.
[   29.385704] H systemd[1]: Started session-14.scope.
[   29.395198] H (o-bridge)[907]: pam_unix(login:session): session opened for user testuser(uid=4711) by (uid=0)
[   29.407198] H dbus-broker[777]: Peer :1.14 is being disconnected as it sent a message with an invalid header.
```

There are three possible issues:

```
[   29.246102] H dbus-broker[777]: Peer :1.12 is being disconnected as it sent a message with an invalid header.
[   29.250866] H dbus-broker[650]: A security policy denied :1.52 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
...
[   29.246429] H (sd-pam)[894]: pam_systemd(login:session): Failed to release session: Access denied
...
[   29.249949] H (sd-pam)[894]: PAM Attempted to close sd-bus after fork, this should not happen.
```

Are any of these something _we_ should/can fix on our side?

_Originally posted by @mrc0mmand in https://github.com/systemd/systemd/issues/28504#issuecomment-1648542218_

**Comment by mrc0mmand** [comment]
The only difference between current main and v253 is the `Attempted to close sd-bus after fork` message, i.e. with v253 I get:

```
Jul 25 17:45:36 bookworm systemd[9770]: Started dbus-broker.service.
Jul 25 17:45:36 bookworm dbus-broker-launch[9778]: Ready
Jul 25 17:45:36 bookworm systemd[9770]: Started run-u0.service.
Jul 25 17:45:36 bookworm dbus-broker[9779]: Peer :1.0 is being disconnected as it sent a message with an invalid header.
Jul 25 17:45:36 bookworm systemd[1]: run-u45.service: Deactivated successfully.
Jul 25 17:45:36 bookworm (sd-pam)[9777]: pam_unix(login:session): session closed for user testuser
Jul 25 17:45:36 bookworm audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=unconfined msg='unit=run-u45 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? r>
Jul 25 17:45:36 bookworm (sd-pam)[9777]: pam_systemd(login:session): Failed to release session: Access denied
Jul 25 17:45:36 bookworm dbus-broker[509]: A security policy denied :1.50 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
Jul 25 17:45:36 bookworm kernel: kauditd_printk_skb: 31 callbacks suppressed
```

Similarly with v252 and v251 (v250 and lower won't build without additional patches):

```
Jul 25 17:47:57 bookworm systemd[527]: Started dbus-broker.service.
Jul 25 17:47:57 bookworm dbus-broker-launch[535]: Ready
Jul 25 17:47:57 bookworm systemd[527]: Started run-u0.service.
Jul 25 17:47:57 bookworm dbus-broker[536]: Peer :1.0 is being disconnected as it sent a message with an invalid header.
Jul 25 17:47:57 bookworm systemd[1]: run-u10.service: Deactivated successfully.
Jul 25 17:47:57 bookworm systemd[534]: pam_unix(login:session): session closed for user testuser
Jul 25 17:47:57 bookworm audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=unconfined msg='unit=run-u10 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? r>
Jul 25 17:47:57 bookworm dbus-broker[445]: A security policy denied :1.15 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
Jul 25 17:47:57 bookworm systemd[534]: pam_systemd(login:session): Failed to release session: Access denied
Jul 25 17:47:57 bookworm systemd[1]: session-3.scope: Deactivated successfully.
```

And as @bluca predicted, the bisect points to the commit that added that warning: 4a75704b166de533cedf8f9fab16ffae77bf2093 (+ the followup ba8d00e859c82b8c1f31d515a48d9b56af1dc9ec that lowers the message to warning).

**Comment by bluca** [comment]
Ok, so there is some deeper but older issue with the ReleaseSession denials. I'll lower the severity of the new warning for now, then it can be bumped up when we solve the ReleaseSession issue.

**Comment by mrc0mmand** [comment]
Huh. I spent some time debugging what exactly is the message with invalid header, and after resorting to strace I think we have a winner:

```strace
clock_gettime(CLOCK_THREAD_CPUTIME_ID, {tv_sec=0, tv_nsec=15004674}) = 0
epoll_wait(3<anon_inode:[eventpoll]>, [], 6, 0) = 0
recvmsg(7<socket:[27719]>, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="l\4\1\1\0\0\0\0\377\377\377\377e\0\0\0\1\1o\0\33\0\0\0/org/freedesktop/DBus/Local\0\0\0\0\0\2\1s\0\32\0\0\0org.freedesktop.DBus.Local\0\0\0\0\0\0\3\1s\0\f\0\0\0Disconnected\0\0\0\0", iov_len=2048}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 120
sendto(4<socket:[22608]>, "PRIORITY=4\nSYSLOG_FACILITY=3\nSYSLOG_IDENTIFIER=dbus-broker\nERRNO=0\nCODE_FILE=../src/bus/peer.c\nCODE_LINE=92\nCODE_FUNC=peer_dispatch_connection\nDBUS_BROKER_LOG_DROPPED=0\nMESSAGE_ID=b209c0d9d1764ab38d13b8e00d1784d6\nDBUS_BROKER_SENDER_SECURITY_LABEL=\nDBUS_BROKER_SENDER_UNIQUE_NAME=:1.19\nMESSAGE=Peer :1.19 is being disconnected as it sent a message with an invalid header.\n", 371, MSG_DONTWAIT|MSG_NOSIGNAL, NULL, 0) = 371
shutdown(7<socket:[27719]>, SHUT_WR)    = 0
```

The culprit seems to be the synthetic `Disconnected` signal:
https://github.com/systemd/systemd/blob/6636b2b6bf07def544dcb4faefe161d27a532692/src/libsystemd/sd-bus/sd-bus.c#L3188-L3197

I see that dbus-daemon has a handler for this: https://gitlab.freedesktop.org/dbus/dbus/-/blob/master/bus/dispatch.c#L326 but I couldn't find anything similar in dbus-broker. The closest thing I was able to get is handling of the `org.freedesktop.DBus.Local` interface: https://github.com/bus1/dbus-broker/blob/156e16c0af846adaabc86849fa8b31b1467d0a00/src/dbus/message.c#L234-L252

Also, just to confirm this, here's journal from `systemd-run --user -M testuser@ true` with both dbus-daemon and dbus-broker:

Arch + dbus-daemon:
```
[   83.192610] archlinux systemd[1]: Started User Manager for UID 1001.
[   83.192650] archlinux systemd[1]: Started Session 5 of User testuser.
[   83.192705] archlinux systemd[630]: Reached target Main User Target.
[   83.192737] archlinux systemd[630]: Startup finished in 105ms.
[   83.192766] archlinux systemd[630]: Created slice User Core Session Slice.
[   83.192795] archlinux systemd[630]: Starting D-Bus User Message Bus...
[   83.195678] archlinux systemd[630]: Started D-Bus User Message Bus.
[   83.197910] archlinux dbus-daemon[638]: [session uid=1001 pid=638] Successfully activated service 'org.freedesktop.systemd1'
[   83.199651] archlinux systemd[630]: Started true.
[   83.201401] archlinux systemd[1]: run-u21.service: Deactivated successfully.
[   83.201687] archlinux (sd-pam)[637]: pam_unix(login:session): session closed for user testuser
[   83.202134] archlinux dbus-daemon[273]: [system] Rejected send message, 2 matched rules; type="method_call", sender=":1.27" (uid=1001 pid=637 comm="(sd-pam)") interface="org.freedesktop.login1.Manager" member="ReleaseSession" error name="(unset)" requested_reply="0" destination="org.freedesktop.login1" (uid=0 pid=274 comm="/usr/lib/systemd/systemd-logind")
[   83.202351] archlinux (sd-pam)[637]: pam_systemd(login:session): Failed to release session: Access denied
```

Arch + dbus-broker
```
[  220.476010] archlinux systemd[408]: Created slice User Core Session Slice.
[  220.477143] archlinux systemd[408]: Starting D-Bus User Message Bus...
[  220.479255] archlinux dbus-broker-launch[429]: Policy to allow eavesdropping in /usr/share/dbus-1/session.conf +31: Eavesdropping is deprecated and ignored
[  220.479255] archlinux dbus-broker-launch[429]: Policy to allow eavesdropping in /usr/share/dbus-1/session.conf +33: Eavesdropping is deprecated and ignored
[  220.479721] archlinux systemd[408]: Started D-Bus User Message Bus.
[  220.480520] archlinux dbus-broker-launch[429]: Ready
[  220.483753] archlinux systemd[408]: Started true.
[  220.484933] archlinux dbus-broker[430]: Peer :1.0 is being disconnected as it sent a message with an invalid header.
[  220.485333] archlinux systemd[1]: run-u44.service: Deactivated successfully.
[  220.485769] archlinux (sd-pam)[428]: pam_unix(login:session): session closed for user testuser
[  220.486040] archlinux dbus-broker[271]: A security policy denied :1.47 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
[  220.486236] archlinux (sd-pam)[428]: pam_systemd(login:session): Failed to release session: Access denied
```

tl;dr the `is being disconnected as it sent a message with an invalid header` message is most likely harmless (as we're about to leave the bus when we get kicked out), but it should get addressed nonetheless.

**Comment by bluca** [comment]
> Huh. I spent some time debugging what exactly is the message with invalid header, and after resorting to strace I think we have a winner:
> 
> ```
> clock_gettime(CLOCK_THREAD_CPUTIME_ID, {tv_sec=0, tv_nsec=15004674}) = 0
> epoll_wait(3<anon_inode:[eventpoll]>, [], 6, 0) = 0
> recvmsg(7<socket:[27719]>, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="l\4\1\1\0\0\0\0\377\377\377\377e\0\0\0\1\1o\0\33\0\0\0/org/freedesktop/DBus/Local\0\0\0\0\0\2\1s\0\32\0\0\0org.freedesktop.DBus.Local\0\0\0\0\0\0\3\1s\0\f\0\0\0Disconnected\0\0\0\0", iov_len=2048}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 120
> sendto(4<socket:[22608]>, "PRIORITY=4\nSYSLOG_FACILITY=3\nSYSLOG_IDENTIFIER=dbus-broker\nERRNO=0\nCODE_FILE=../src/bus/peer.c\nCODE_LINE=92\nCODE_FUNC=peer_dispatch_connection\nDBUS_BROKER_LOG_DROPPED=0\nMESSAGE_ID=b209c0d9d1764ab38d13b8e00d1784d6\nDBUS_BROKER_SENDER_SECURITY_LABEL=\nDBUS_BROKER_SENDER_UNIQUE_NAME=:1.19\nMESSAGE=Peer :1.19 is being disconnected as it sent a message with an invalid header.\n", 371, MSG_DONTWAIT|MSG_NOSIGNAL, NULL, 0) = 371
> shutdown(7<socket:[27719]>, SHUT_WR)    = 0
> ```
> 
> The culprit seems to be the synthetic `Disconnected` signal:
> 
> https://github.com/systemd/systemd/blob/6636b2b6bf07def544dcb4faefe161d27a532692/src/libsystemd/sd-bus/sd-bus.c#L3188-L3197
> 
> I see that dbus-daemon has a handler for this: https://gitlab.freedesktop.org/dbus/dbus/-/blob/master/bus/dispatch.c#L326 but I couldn't find anything similar in dbus-broker. The closest thing I was able to get is handling of the `org.freedesktop.DBus.Local` interface: https://github.com/bus1/dbus-broker/blob/156e16c0af846adaabc86849fa8b31b1467d0a00/src/dbus/message.c#L234-L252
> 
> Also, just to confirm this, here's journal from `systemd-run --user -M testuser@ true` with both dbus-daemon and dbus-broker:
> 
> Arch + dbus-daemon:
> 
> ```
> [   83.192610] archlinux systemd[1]: Started User Manager for UID 1001.
> [   83.192650] archlinux systemd[1]: Started Session 5 of User testuser.
> [   83.192705] archlinux systemd[630]: Reached target Main User Target.
> [   83.192737] archlinux systemd[630]: Startup finished in 105ms.
> [   83.192766] archlinux systemd[630]: Created slice User Core Session Slice.
> [   83.192795] archlinux systemd[630]: Starting D-Bus User Message Bus...
> [   83.195678] archlinux systemd[630]: Started D-Bus User Message Bus.
> [   83.197910] archlinux dbus-daemon[638]: [session uid=1001 pid=638] Successfully activated service 'org.freedesktop.systemd1'
> [   83.199651] archlinux systemd[630]: Started true.
> [   83.201401] archlinux systemd[1]: run-u21.service: Deactivated successfully.
> [   83.201687] archlinux (sd-pam)[637]: pam_unix(login:session): session closed for user testuser
> [   83.202134] archlinux dbus-daemon[273]: [system] Rejected send message, 2 matched rules; type="method_call", sender=":1.27" (uid=1001 pid=637 comm="(sd-pam)") interface="org.freedesktop.login1.Manager" member="ReleaseSession" error name="(unset)" requested_reply="0" destination="org.freedesktop.login1" (uid=0 pid=274 comm="/usr/lib/systemd/systemd-logind")
> [   83.202351] archlinux (sd-pam)[637]: pam_systemd(login:session): Failed to release session: Access denied
> ```
> 
> Arch + dbus-broker
> 
> ```
> [  220.476010] archlinux systemd[408]: Created slice User Core Session Slice.
> [  220.477143] archlinux systemd[408]: Starting D-Bus User Message Bus...
> [  220.479255] archlinux dbus-broker-launch[429]: Policy to allow eavesdropping in /usr/share/dbus-1/session.conf +31: Eavesdropping is deprecated and ignored
> [  220.479255] archlinux dbus-broker-launch[429]: Policy to allow eavesdropping in /usr/share/dbus-1/session.conf +33: Eavesdropping is deprecated and ignored
> [  220.479721] archlinux systemd[408]: Started D-Bus User Message Bus.
> [  220.480520] archlinux dbus-broker-launch[429]: Ready
> [  220.483753] archlinux systemd[408]: Started true.
> [  220.484933] archlinux dbus-broker[430]: Peer :1.0 is being disconnected as it sent a message with an invalid header.
> [  220.485333] archlinux systemd[1]: run-u44.service: Deactivated successfully.
> [  220.485769] archlinux (sd-pam)[428]: pam_unix(login:session): session closed for user testuser
> [  220.486040] archlinux dbus-broker[271]: A security policy denied :1.47 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
> [  220.486236] archlinux (sd-pam)[428]: pam_systemd(login:session): Failed to release session: Access denied
> ```
> 
> tl;dr the `is being disconnected as it sent a message with an invalid header` message is most likely harmless (as we're about to leave the bus when we get kicked out anyway), but it should get addressed nonetheless.

Good find - @dvdhrm  is this Disconnect signal something that could be handled by dbus-broker too?

**Comment by mrc0mmand** [comment]
And just to confirm this even further, it can be easily reproduced with `busctl`:

```
$ journalctl --cursor-file=cursor -n0 -q
$ busctl emit /org/freedesktop/DBus/Local org.freedesktop.DBus.Local Connected
$ busctl emit /org/freedesktop/DBus/Local org.freedesktop.DBus.Local Disconnected
$ journalctl --cursor-file=cursor -o short-monotonic --no-hostname
[215484.889033] dbus-broker[1565]: Peer :1.6257 is being disconnected as it sent a message with an invalid header.
[215487.446986] dbus-broker[1565]: Peer :1.6258 is being disconnected as it sent a message with an invalid header.
```

The same commands with dbus-daemon work fine (or at least dbus-daemon doesn't complain).

**Comment by bluca** [comment]
While we wait on an answer on the Disconnected, going back to the other issue:

```
[   29.303401] H (o-bridge)[900]: pam_unix(login:session): session opened for user testuser(uid=4711) by (uid=0)
[   29.323777] H (sd-pam)[901]: pam_systemd(login:session): Failed to release session: Access denied
[   29.325619] H (sd-pam)[901]: pam_unix(login:session): session closed for user testuser
[   29.325743] H (sd-pam)[901]: PAM Attempted to close sd-bus after fork, this should not happen.
```

Does this by any chance mean we are opening a session from one process, but then trying to close it from another? Is that even allowed?

**Comment by mrc0mmand** [comment]
> While we wait on an answer on the Disconnected, going back to the other issue:
> 
> ```
> [   29.303401] H (o-bridge)[900]: pam_unix(login:session): session opened for user testuser(uid=4711) by (uid=0)
> [   29.323777] H (sd-pam)[901]: pam_systemd(login:session): Failed to release session: Access denied
> [   29.325619] H (sd-pam)[901]: pam_unix(login:session): session closed for user testuser
> [   29.325743] H (sd-pam)[901]: PAM Attempted to close sd-bus after fork, this should not happen.
> ```
> 
> Does this by any chance mean we are opening a session from one process, but then trying to close it from another? Is that even allowed?

No idea, unfortunately. All I can say is that the dbus policy message is not related to this, since even after adding a policy for the ReleaseSession method the access denied error remains.

**Comment by dvdhrm** [comment]
> Good find - @dvdhrm is this Disconnect signal something that could be handled by dbus-broker too?

I don't think so. The spec says: (https://dbus.freedesktop.org/doc/dbus-specification.html)

Object-paths:
> The object to send a call to, or the object a signal is emitted from. The special path /org/freedesktop/DBus/Local is reserved; implementations should not send messages with this path, and the reference implementation of the bus daemon will disconnect any application that attempts to do so. This header field is controlled by the message sender.

Interfaces:
> The interface to invoke a method call on, or that a signal is emitted from. Optional for method calls, required for signals. The special interface org.freedesktop.DBus.Local is reserved; implementations should not send messages with this interface, and the reference implementation of the bus daemon will disconnect any application that attempts to do so. This header field is controlled by the message sender.

Hello():
>  There is no corresponding "disconnect" request; if a client wishes to disconnect from the bus, it simply closes the socket (or other communication channel).

Also, I think dbus-daemon and dbus-broker behave the same for this particular signal, don't they? They simply disconnect the client. It is just that dbus-broker complains that you violated the spec, dbus-daemon does not, right?

**Comment by bluca** [comment]
I see - @smcv do you happen to remember why dbus-daemon lets clients send a Disconnected signal?

**Comment by smcv** [comment]
> do you happen to remember why dbus-daemon lets clients send a Disconnected signal?

It ... doesn't? `load_and_validate_field()` considers any method with interface `org.freedesktop.DBus.Local` to be invalid.

```
$ gdbus call --session -o / -d com.example.Nope -m org.freedesktop.DBus.Local.Nope
Error: The connection is closed
```

I think the only difference is that dbus-broker logs `Peer :1.0 is being disconnected as it [did something bad]`, whereas dbus-daemon just silently disconnects you. Clients should never send messages with this reserved interface or the reserved path `/org/freedesktop/DBus/Local`, and dbus-daemon checks for both.

The Disconnected signal on the reserved local path/interface is never meant to be sent or received over the socket: it's purely an API convention for client libraries, so that clients can handle being disconnected via the code they already needed to have to receive "real" signals, and client libraries don't need to have a separate mechanism to report "I was disconnected" to the application's event loop. (IMO that's a bit of a hack, but it was designed before my involvement.)

dbus-specification only reserves the interface and path to prevent them from being used for something else, and doesn't describe the Disconnected signal itself, because the Disconnected signal is really more like part of the libdbus API than part of the D-Bus wire protocol; but other D-Bus implementations like GDBus and sd-bus have presumably copied libdbus' use of the Disconnected signal as a semi-convenient way to report disconnection.

**Comment by bluca** [comment]
Alright, this sounds like a red herring then, thanks for the explanation

**Comment by mrc0mmand** [comment]
> While we wait on an answer on the Disconnected, going back to the other issue:
> 
> ```
> [   29.303401] H (o-bridge)[900]: pam_unix(login:session): session opened for user testuser(uid=4711) by (uid=0)
> [   29.323777] H (sd-pam)[901]: pam_systemd(login:session): Failed to release session: Access denied
> [   29.325619] H (sd-pam)[901]: pam_unix(login:session): session closed for user testuser
> [   29.325743] H (sd-pam)[901]: PAM Attempted to close sd-bus after fork, this should not happen.
> ```
> 
> Does this by any chance mean we are opening a session from one process, but then trying to close it from another? Is that even allowed?

From a quick debug session the main issue is that both `CreateSession` and `ReleaseSession` are privileged methods (according to the `org.freedesktop.login1.Manager.xml` interface spec), but we're running with different EUID in `pam_sm_close_session()`:

```
# systemd-run --user -M testuser@ true
pam_sm_open_session: EUID: 0
pam_sm_open_session: Created session 12
Running as unit: run-u0.service
pam_sm_close_session: EUID: 1001
pam_sm_close_session: About to release session 12
```

**Comment by mrc0mmand** [comment]
Anyway, this issue has been there for a long time (I could reproduce it with cbdc29492097e24ef3320280bc2a8dedbce02d9a which is iiuc the first commit that makes the reproducer I use possible), so it's nothing urgent. But it should get addressed in the next release, as I'm pretty sure folks will complain, especially since it logs at the error log level (and we already got some reports like https://bugzilla.redhat.com/show_bug.cgi?id=2221337).

**Comment by bluca** [comment]
@poettering did you mention you know how to fix this issue?

**Comment by poettering** [comment]
Fix for the org.freedesktop.DBus.Local.Disconnected message waits in #30242.

**Comment by poettering** [comment]
#30242 only addresse done of the log messages, the other one (about the perms to access ReleaseSession()) is not addressed. Reopening hence.

**Comment by poettering** [comment]
So let's move this one to the next milestone. It's mostly cosmetical only. it's not hard to fix, but I don't want to touch this code before the reworks in #30226 are merged really.

Basically, here's what's going on:

Whenever PAM is enabled for a service, we set up the PAM session and then fork off a process whose only job is to eventually close the PAM session when the service dies. That services we run with service privileges, both to minimize attack surface and because we want to use PR_SET_DEATHSIG to be get a notification via signal whenever the main process dies. But that only works if we have the same credentials as that main process.

Now, if pam_systemd runs inside the PAM stack (which it normally does) it's session close hook will ask logind to synchronously end the session via a bus call. Currently that call is not accessible to unprivileged clients. And that's the part we need to relax: allow users to end their own sessions.

This is mostly cosmetical, since if we can't close our session we'll simply die which also closes the session, and in this specific case we are operating async anyway (as reaction to PR_SET_DEATHSIG signal), hence the benefit of synchronous behaviour doesn't get us anything.

Hence: yes, we should fix this, to make dbus-broker's log messages go away, but also doesn#t matter for now.

**Comment by mrc0mmand** [comment]
Also related: https://github.com/systemd/systemd/issues/8598

**Comment by tijko** [comment]
> Also related: #8598

@mrc0mmand I have been tracking another case where this is popping up.

Running `systemd-255.4-1-arch` with `Linux tijko 6.7.6-arch1-1 #1 SMP PREEMPT_DYNAMIC Fri, 23 Feb 2024 16:31:48 +0000 x86_64 GNU/Linux`

Not to cause any more noise on this thread and I'm not wanting to be disrespectful of your time and attention.  Just thought I could be pointed in the right direction from here.
